funky-flamingos/rinky_dink/users/views.py

```diff
@@ -0,0 +1,199 @@
+    return render(request, "users/login.html")
+
+
+@login_required(login_url="/login/")
+def upload(request):
+    context = {}
+    if request.method == "POST":
+        uploaded_file = request.FILES["document"]
+        if not check_title_presence(uploaded_file):
+            messages.success(request, "New file save successfully.")
+            path = save_on_server(uploaded_file)
+            hash_val = hash_file(path)
+            form = File(title=uploaded_file.name, hash_val=hash_val, user=request.user)
+            form.save()
+            file_group_form = FileGroup(title=uploaded_file.name, user=request.user)
+            file_group_form.save()
+            file_group_form.files.add(form)
+        else:
+            # flash message that A similar file is already present  if hash value matches
+            path = save_on_server(uploaded_file)
+            hash_val = hash_file(path)
+            if check_file_presence(hash_val):
+                messages.warning(request, "A similar file is already present")
+                os.remove(path)
+
+            # hash value matches so no need to upload. if not, upload it with different file name
+            else:
+                file_group = FileGroup.objects.filter(title=uploaded_file.name)[0]
+                # save file on server with new name{version}
+                title = path.split("/")[-1]
+                form = File(title=title, hash_val=hash_val, user=request.user)
+                form.save()
+                file_group.files.add(
+                    form
+                )  # adds newly created file version to file group
+                messages.warning(
+                    request,
+                    "file with same name is already present adding it with new version ",
+                )
+
+    return render(request, "users/upload.html", context)
+
+
+@login_required(login_url="/login/")
+def dashboard(request):
+    teams = Member.objects.filter(user__id__in=[request.user.id])
+    context = {}
+    files = FileGroup.objects.filter(user__id__in=[request.user.id])
+    context = {"files": files, "teams": teams}
+    print(context["teams"])
+    return render(request, "users/dashboard.html", context)
+
+
+def logout(request):
+    auth.logout(request)
+    messages.success(request, "You have been logged out")
+    return render(request, "users/logout.html")
+
+
+def hash_file(filename):
+    """"This function returns the SHA-1 hash
+    of the file passed into it"""
+
+    # make a hash object
+    h = hashlib.sha1()
+    with open(filename, "rb") as file:
+
+        # loop till the end of the file
+        chunk = 0
+        while chunk != b"":
+            # read only 1024 bytes at a time
+            chunk = file.read(1024)
+            h.update(chunk)
+
+    # return the hex representation of digest
+    return h.hexdigest()
+
+
+def check_title_presence(uploaded_file):
+    """
+    query database for filename and if it is present then return True else return False
```

Comment:
to comply with PEP257, docstrings should start with capital letters and end with punctuation. In the docstring, don't explain return types (this can be explained with return type annotations), and don't explain parameters (these can be type annotated). Instead, describe at a high level what action the function performs. e.g., here `Check whether a specific file is present in the database.` would've been fine.